OpenRA.Platforms.Default/OpenAlSoundEngine.cs

```diff
@@ -404,54 +412,256 @@ public void Dispose()
+					if (state == AL10.AL_INITIAL)
+						AL10.alSourcePlay(Source);
+
+					// Ensure the source is stopped, which will mark all buffers as processed.
+					// Dequeue these, so they can then be freed.
+					AL10.alSourceStop(Source);
+					lock (bufferDequeueLock)
+					{
+						int buffersProcessed;
+						AL10.alGetSourcei(source, AL10.AL_BUFFERS_PROCESSED, out buffersProcessed);
+						for (var i = 0; i < buffersProcessed; i++)
+						{
+							var dequeued = 0U;
+							AL10.alSourceUnqueueBuffers(source, new IntPtr(1), ref dequeued);
+						}
+					}
+
+					AL10.alDeleteBuffers(new IntPtr(buffers.Length), buffers);
+					stream.Dispose();
+				}
+			}).ContinueWith(task =>
+			{
+				Game.RunAfterTick(() =>
+				{
+					throw new Exception("Failed to stream a sound.", task.Exception);
+				});
+			}, TaskContinuationOptions.OnlyOnFaulted);
+		}
+
+		async Task<int> ReadFillingBuffer(Stream stream, byte[] buffer)
+		{
+			var offset = 0;
+			int count;
+			while (offset < buffer.Length &&
+				(count = await stream.ReadAsync(buffer, offset, buffer.Length - offset, cts.Token).ConfigureAwait(false)) > 0)
+				offset += count;
+			return offset;
+		}
+
+		void DequeueBuffers(uint source, Stack<uint> freeBuffers)
+		{
+			lock (bufferDequeueLock)
+			{
+				// Check for any processed buffers, and dequeue them for reuse.
+				int buffersProcessed;
+				AL10.alGetSourcei(source, AL10.AL_BUFFERS_PROCESSED, out buffersProcessed);
+				for (var i = 0; i < buffersProcessed; i++)
+				{
+					// Dequeue a processed buffer, so we can reuse it.
+					var dequeued = 0U;
+					AL10.alSourceUnqueueBuffers(source, new IntPtr(1), ref dequeued);
+					freeBuffers.Push(dequeued);
+
+					// When we remove a buffer, we need to account for this to calculate the overall seek position.
+					totalBuffersPlayed++;
+				}
+			}
+		}
+
+		public override void Stop()
+		{
+			lock (cts)
+			{
+				StopSource();
+				cts.Cancel();
+			}
+
+			try
+			{
+				streamTask.Wait();
+			}
+			catch (AggregateException)
+			{
+			}
+		}
+
+		public override float SeekPosition
+		{
+			get
+			{
+				// Stop buffers being dequeued whilst we calculate the seek position.
+				lock (bufferDequeueLock)
+				{
+					int sampleOffset;
+					AL10.alGetSourcei(Source, AL11.AL_SAMPLE_OFFSET, out sampleOffset);
+
+					int state;
+					AL10.alGetSourcei(Source, AL10.AL_SOURCE_STATE, out state);
+
+					// If the source is not stopped, add the current offset to the total offset and return that.
+					if (state != AL10.AL_STOPPED)
+						return sampleOffset / SampleRate + totalBuffersPlayed * BufferSizeInSecs;
+
+					// If the source stopped, the current offset will have been reset to 0.
+					// Instead, add the length of processed buffers to the total and return that.
+					int buffersProcessed;
+					AL10.alGetSourcei(Source, AL10.AL_BUFFERS_PROCESSED, out buffersProcessed);
+
+					return (totalBuffersPlayed + buffersProcessed) * BufferSizeInSecs;
```

Comment:
The last buffer of the track might not be one full second long. Accounting for the actual number of samples in it fixes the accounting (I was doing this originally, should've commented it so I didn't forget why!)